pkg/resource/graph/resource_set.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package graph
+
+import "github.com/pulumi/pulumi/pkg/resource"
+
+// ResourceSet represents a set of Resources.
+type ResourceSet struct {
```

Comment:
This seems like overkill--why not just use a `map` directly?

Reply:
I wanted `ResourceSet.Intersect` and it's annoying to do that with just maps. I don't mind having this either way, I was just not a fan of doing set intersections of with `map[*resource.State]bool`.

Reply:
Note that you can just do

```
type ResourceSet map[*resource.State]bool

func (rs ResourceSet) Intersect(other ResourceSet) ResourceSet {
    i := make(ResourceSet)
    for r := range rs {
        if other[r] {
            i[r] = true
        }
    }
    return i
}
```

And then just interact with the set normally otherwise.

Reply:
Oh, cool, I wasn't aware that you could do this while also preserving map-like behavior.